src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,234 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdint>
+#include <string>
+
+#include "absl/base/internal/endian.h"
+#include "absl/strings/string_view.h"
+#include "common/buffer/buffer_impl.h"
+#include "envoy/network/connection.h"
+#include "envoy/stats/scope.h"
+#include "src/envoy/tcp/alpn_proxy/alpn_proxy.h"
+#include "src/envoy/tcp/alpn_proxy/alpn_proxy_initial_header.h"
+#include "src/envoy/tcp/alpn_proxy/header/alpn_proxy_header.pb.h"
+
+namespace Envoy {
+namespace Tcp {
+namespace AlpnProxy {
+namespace {
+
+std::unique_ptr<::Envoy::Buffer::OwnedImpl> constructProxyHeaderData(
+    const ::google::protobuf::Struct& metadata) {
+  envoy::tcp::alpnproxy::header::AlpnProxyHeaderProto proxy_header;
+  proxy_header.mutable_metadata()->CopyFrom(metadata);
+
+  AlpnProxyInitialHeader initial_header;
+  initial_header.magic = absl::ghtonl(AlpnProxyInitialHeader::magic_number);
+  initial_header.data_size =
+      absl::ghtonl(proxy_header.SerializeAsString().length());
```

Comment:
Please serialize to string once and save the string on stack.

Reply:
Done